pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -521,45 +605,61 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 	// Note that we do not resurrect Pods in this interval. Also note that scaling will take precedence over
 	// updates.
 	for target := len(condemned) - 1; target >= 0; target-- {
+		target := target
 		// wait for terminating pods to expire
 		if isTerminating(condemned[target]) {
 			logger.V(4).Info("StatefulSet is waiting for Pod to Terminate prior to scale down",
 				"statefulSet", klog.KObj(set), "pod", klog.KObj(condemned[target]))
 			// block if we are in monotonic mode
 			if monotonic {
-				return &status, nil
+				break
 			}
 			continue
 		}
 		// if we are in monotonic mode and the condemned target is not the first unhealthy Pod block
 		if !isRunningAndReady(condemned[target]) && monotonic && condemned[target] != firstUnhealthyPod {
 			logger.V(4).Info("StatefulSet is waiting for Pod to be Running and Ready prior to scale down",
 				"statefulSet", klog.KObj(set), "pod", klog.KObj(firstUnhealthyPod))
-			return &status, nil
+			break
 		}
 		// if we are in monotonic mode and the condemned target is not the first unhealthy Pod, block.
 		if !isRunningAndAvailable(condemned[target], set.Spec.MinReadySeconds) && monotonic && condemned[target] != firstUnhealthyPod {
 			logger.V(4).Info("StatefulSet is waiting for Pod to be Available prior to scale down",
 				"statefulSet", klog.KObj(set), "pod", klog.KObj(firstUnhealthyPod))
-			return &status, nil
+			break
 		}
-		logger.V(2).Info("Pod of StatefulSet is terminating for scale down",
-			"statefulSet", klog.KObj(set), "pod", klog.KObj(condemned[target]))
 
-		if err := ssc.podControl.DeleteStatefulPod(set, condemned[target]); err != nil {
-			return &status, err
-		}
-		if getPodRevision(condemned[target]) == currentRevision.Name {
-			status.CurrentReplicas--
+		actions = append(actions, func() (bool, error) {
+			logger.V(2).Info("Pod of StatefulSet is terminating for scale down",
+				"statefulSet", klog.KObj(set), "pod", klog.KObj(condemned[target]))
+			return true, ssc.podControl.DeleteStatefulPod(set, condemned[target])
+		})
+
+		if monotonic {
+			break
 		}
-		if getPodRevision(condemned[target]) == updateRevision.Name {
-			status.UpdatedReplicas--
+	}
+
+	if monotonic {
+		for _, fn := range actions {
+			if shouldExit, err := fn(); shouldExit || err != nil {
+				return &status, err
+			}
 		}
-		if monotonic {
-			return &status, nil
+	} else {
+		// Parallelize with slow start.
+		// TODO: limit max batch size/max remaining
+		successes, err := slowStartBatch(1, actions)
+		logger.V(2).Info(fmt.Sprintf("Executed %v successful actions", successes), "statefulSet", klog.KObj(set))
+		if err != nil {
+			updateStatus(&status, set.Spec.MinReadySeconds, currentRevision, updateRevision, replicas, condemned)
+			return &status, err
 		}
 	}
 
+	// Update status so the rolling update logic can rely on it and update it as necessary.
+	updateStatus(&status, set.Spec.MinReadySeconds, currentRevision, updateRevision, replicas, condemned)
+
```

Comment:
You probably need to leave function here if you breaked, e.g. on line 629.

Reply:
Added action returning shouldExit=true in that case